docs/hashing.rst

```diff
@@ -0,0 +1,159 @@
+    b'3235362042797465734d6573736167653235362042797465734d657373616765'
+    b'3235362042797465734d6573736167653235362042797465734d657373616765'
+    b'3235362042797465734d6573736167653235362042797465734d657373616765'
+    b'3235362042797465734d6573736167653235362042797465734d657373616765'
+    b'3235362042797465734d6573736167653235362042797465734d657373616765'
+    b'3235362042797465734d6573736167653235362042797465734d657373616765'
+    b'3235362042797465734d6573736167653235362042797465734d657373616765'
+    )
+
+    dgst = b'12b413c70c148d79bb57a1542156c5f35e24ad77c38e8c0e776d055e827cdd45'
+
+    shortened = received_msg[:-1]
+    modified = b'modified' + received_msg[:-8]
+
+    orig_dgs = HASHER(received_msg, encoder=nacl.encoding.HexEncoder)
+    shrt_dgs = HASHER(shortened, encoder=nacl.encoding.HexEncoder)
+    mdfd_dgs = HASHER(modified, encoder=nacl.encoding.HexEncoder)
+
+    def eq_chk(dgs0, dgs1):
+        if sodium_memcmp(dgs0, dgs1):
+            return 'equals'
+        return 'is different from'
+
+    MSG = 'Digest of {0} message {1} original digest'
+
+    for chk in (('original', orig_dgs),
+                ('truncated', shrt_dgs),
+                ('modified', mdfd_dgs)):
+
+        print(MSG.format(chk[0], eq_chk(dgst, chk[1])))
+
+
+Additional hashing usages for :class:`~nacl.hash.blake2b`:
+==========================================================
+
+As already hinted above, traditional cryptographic hash functions can be used
+as building blocks for other uses, tipically combining a secret-key with
+the message via some construct like the ``HMAC`` one.
+
+The :class:`~nacl.hash.blake2b` hash function can be used directly both
+for message authentication and key derivation, replacing the ``HMAC`` construct
+and the ``HKDF`` one by setting the additional parameters ``key``, ``salt``
+and ``person``.
+
+Message authentication:
+-----------------------
+
+To authenticate a message, using a secret key, the blake2b function
+must be called as in the following example.
+
+Message authentication example:
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code-block:: python
+
+    import nacl.encoding
+    from nacl.hash import blake2b
+
+    msg = 16*b'256 BytesMessage'
+    msg2 = 16*b'256 bytesMessage'
+
+    auth_key = b''.join([bytearray((i,)) for i in range(64)])
+    # should really be a cryptographic quality random key, like
+    # the ones you could get with nacl.utils.random(size=64)
+
+    auth1_key = b''.join([bytearray((i,)) for i in range(64,128)])
+
+    mac0 = blake2b(msg, key=auth_key, encoder=nacl.encoding.HexEncoder)
+    mac1 = blake2b(msg, key=auth1_key, encoder=nacl.encoding.HexEncoder)
+    mac2 = blake2b(msg2, key=auth_key, encoder=nacl.encoding.HexEncoder)
+
+    for i, mac in enumerate((mac0, mac1, mac2)):
+        print('Mac{0} is: {1}.'.format(i, mac))
+
+
+Key derivation:
+---------------
+
+The blake2b algorithm can replace a key derivation function by
```

Comment:
@reaperhulk I think 4b8428e should cover both your review comments for now. I'm trying to write a page on password hashing, but I would rather open a separate pull request for that.